### PR TITLE
Update the MockOptimizer set_to_none default to match PyTorch

### PR DESCRIPTION
**Pull request by janeyx99**
## What does this PR do?

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/pytorch/core/optimizer.py [hunk 1]
```diff
@@ -382,7 +382,7 @@ def step(self, closure: Optional[Callable] = None) -> None:
         if closure is not None:
             closure()
 
-    def zero_grad(self, set_to_none: Optional[bool] = False) -> None:
+    def zero_grad(self, set_to_none: Optional[bool] = True) -> None:
```
Comment (carmocca) [inline src/lightning/pytorch/core/optimizer.py:385]: Even though it's fine to update this, it won't have any performance implication. This class is just a mock we use when `configure_optimizers` is not defined (generally because the Trainer is evaluating)

Lightning just uses the optimizer returned from `configure_optimizers` when training

Comment (janeyx99) [reply src/lightning/pytorch/core/optimizer.py:385]: Thanks for the quick response! Does this mean lightning already defaults to set_to_none=True? If so, that's great and this PR isn't quite necessary.

**Review by carmocca — APPROVED** [review]

File: src/lightning/pytorch/core/optimizer.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/pytorch/core/optimizer.py:382]: I'm also fine with merging this.

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/core/optimizer.py (hunk 1 above)
Comment (janeyx99) [reply src/lightning/pytorch/core/optimizer.py:382]: Sounds good, I'll clean up the PR description to say so. Thanks for your input everyone!
